src/nupic/algorithms/SDRClassifier.cpp

```diff
@@ -0,0 +1,745 @@
+      {
+        // save the offset between recordNum and learnIteration_ if this
+        // was not set (first call to compute)
+        if (!recordNumMinusLearnIterationSet_)
+        {
+          recordNumMinusLearnIteration_ = recordNum - learnIteration_;
+          recordNumMinusLearnIterationSet_ = true;
+        }
+
+        // update learnIteration_
+        learnIteration_ = recordNum - recordNumMinusLearnIteration_;
+
+        // update pattern history
+        patternNZHistory_.emplace_front(patternNZ.begin(), patternNZ.end());
+        iterationNumHistory_.push_front(learnIteration_);
+        if (patternNZHistory_.size() > maxSteps_)
+        {
+          patternNZHistory_.pop_back();
+          iterationNumHistory_.pop_back();
+        }
+
+        // if input pattern has greater index than previously seen, update 
+        // maxInputIdx and augment weight matrix with zero padding
+        UInt maxInputIdx = *max_element(patternNZ.begin(), patternNZ.end());
+        if (maxInputIdx > maxBucketIdx_)
+        {
+          maxInputIdx_ = maxInputIdx;
+          for (const auto& step : steps_)
+          { 
+            weightMatrix_[step].resize(maxInputIdx_ + 1, maxBucketIdx_ + 1);
+          }
+        }
+
+        // if in inference mode, compute likelihood and update return value
+        if (infer)
+        {
+            infer_(patternNZ, bucketIdx, actValue, result);
+        }
+
+        // update weights if in learning mode
+        if (learn)
+        {
+          // if bucket is greater, updated maxBucketIdx_ and augment weight
+          // matrix with zero-padding
+          if (bucketIdx > maxBucketIdx_) 
+          {
+            maxBucketIdx_ = bucketIdx;
+            for (const auto& step : steps_)
+            {
+              weightMatrix_[step].resize(maxInputIdx_ + 1, maxBucketIdx_ + 1);
+            }
+          }
+
+          // update rolling averages of bucket values
+          while (actualValues_.size() <= maxBucketIdx_)
+          {
+            actualValues_.push_back(0.0);
+            actualValuesSet_.push_back(false);
+          }
+          if (!actualValuesSet_[bucketIdx] || category)
+          {
+            actualValues_[bucketIdx] = actValue;
+            actualValuesSet_[bucketIdx] = true;
+          } else {
+            actualValues_[bucketIdx] =
+                ((1.0 - actValueAlpha_) * actualValues_[bucketIdx]) +
+                (actValueAlpha_ * actValue);
+          }
+
+          // compute errors and update weights
+          deque<vector<UInt>>::const_iterator patternIteration =
+              patternNZHistory_.begin();
+          for (deque<UInt>::const_iterator learnIteration =
+               iterationNumHistory_.begin();
+               learnIteration !=iterationNumHistory_.end();
+               learnIteration++, patternIteration++)
+          {
+            const vector<UInt> learnPatternNZ = *patternIteration;
+            UInt nSteps = learnIteration_ - *learnIteration;
```

Comment:
It helps readability to make stuff `const` when you can, e.g. `nSteps` here. Then I can know at a glance that this value isn't about to be manipulated below.